data_model/src/predicate.rs

```diff
@@ -46,10 +77,45 @@ impl PredicateBox {
             Self::And(preds) => Self::Or(preds.into_iter().map(Self::negate).collect()),
             Self::Or(preds) => Self::And(preds.into_iter().map(Self::negate).collect()),
             Self::Not(pred) => *pred, // TODO: should we recursively simplify?
-            Self::Raw(pred) => Self::Not(Box::new(PredicateBox::Raw(pred))),
+            Self::Raw(pred) => Self::Not(Box::new(Self::Raw(pred))),
         }
     }
+}
+
+impl<Pred, Input> PredicateTrait<Input> for GenericPredicateBox<Pred>
+where
+    Input: ?Sized + Copy,
+    Pred: PredicateTrait<Input>,
+{
+    type EvaluatesTo = Pred::EvaluatesTo;
 
+    #[inline] // This is not a simple function, but it allows you to inline the logic and optimise away the logical operations.
```

Comment:
Good point. Though, you'd have to recognise that it is in fact a simple function once compiled. Most of the text is type noise, just like accessor methods.